Fix/relations override prop collections

kiuby88 opened the pull request:
When a relationship is configured the a `ConfigKey` value (maybe a `Flag`) is added to the EntitySpec`. However, this value addition does not take in account any old config value which might have configured previously. For example, 

```
   tomcat_server:
      type: org.apache.brooklyn.entity.webapp.tomcat.TomcatServer
      properties:
        java.sysprops:
          key1: value1
      requirements:
        - endpoint_configuration:
            node: backend
            relationship: brooklyn.relationships.Configure
            properties:
              prop.collection: java.sysprops
              prop.name: key2
              prop.value: value2
```

The generated `EntitySpec` has to contains two entries in  `java.sysprops` configuration field.
 - `key1`: `value1``
 - `key2`: `value2`

However, currently the generated entity just contains the `key2: value2` entry. 
Currently, `RelationshipModifier` adds a configuration to an spec invoking `ConfigKeyModifier.configureConfigKeysSpec` method,  which replace the previous values of any `EntitySpec`'s configuration. 
It looks that `ConfigKeyModifier.configureConfigKeysSpec` must be used for this task, because this method adds a configuration value to an entity spec, and if the entity spec already contains a  value for this configuration `ConfigKeyModifier.configureConfigKeysSpec` should replace the old value.

sjcorbett commented on brooklyn-tosca-transformer/src/main/java/io/cloudsoft/tosca/a4c/brooklyn/spec/ConfigKeyModifier.java:
```diff
@@ -68,7 +97,66 @@ private void configureWithResolvedConfigKey(FlagUtils.FlagConfigKeyAndValueRecor
+        } else if ((oldValue instanceof String) && (newValue instanceof String)) {
+            return newValue;
+        } else {
+            LOG.warn("Types of oldValue {} and newValue {} are not the same, so {} is not able " +
```
This isn't worth a warning - probably trace, maybe debug. I don't think there's any need to special case String either.

sjcorbett commented on brooklyn-tosca-transformer/src/main/java/io/cloudsoft/tosca/a4c/brooklyn/spec/ConfigKeyModifier.java:
```diff
@@ -68,7 +97,66 @@ private void configureWithResolvedConfigKey(FlagUtils.FlagConfigKeyAndValueRecor
+
+    private Optional<Object> getConfigKeyValue(EntitySpec spec,
+                                               FlagUtils.FlagConfigKeyAndValueRecord r) {
+        Object configValue = spec.getConfig().get(r.getConfigKey());
```
You can just write `return Optional.fromNullable(spec.getConfig().get(r.getConfigKey()))`

sjcorbett commented on brooklyn-tosca-transformer/src/main/java/io/cloudsoft/tosca/a4c/brooklyn/spec/ConfigKeyModifier.java:
```diff
@@ -68,7 +97,66 @@ private void configureWithResolvedConfigKey(FlagUtils.FlagConfigKeyAndValueRecor
+
+    private Optional<Object> findFlagValue(EntitySpec spec,
+                                           FlagUtils.FlagConfigKeyAndValueRecord r) {
+        Object flagValue = spec.getFlags().get(r.getFlagName());
```
And here.

sjcorbett commented on brooklyn-tosca-transformer/src/test/java/io/cloudsoft/tosca/a4c/brooklyn/ToscaTypePlanTransformerIntegrationTest.java:
```diff
@@ -34,25 +27,25 @@
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertNotNull;
+import static org.testng.Assert.assertTrue;
```
Please configure your IDE to not automatically sort its imports. Or alternatively, configure it to use Eclipse's order. There are instructions for IntelliJ at http://stackoverflow.com/a/17194980/41195.

kiuby88 replied:
Sorry, for that. I have just added to my IDE the Eclipse's order

sjcorbett commented on brooklyn-tosca-transformer/src/test/java/io/cloudsoft/tosca/a4c/brooklyn/ToscaTypePlanTransformerIntegrationTest.java:
```diff
@@ -189,6 +182,51 @@ public void testRelation() throws ParsingException, CSARVersionAlreadyExistsExce
 
+
+    @Test
+    public void testRelationNotOverridePropCollectionConfigKey()
```
These tests are good but in general prefer adding unit tests with mocked dependencies because they're _much_ quicker to run repeatedly than these integration tests that require a full Spring context.

sjcorbett commented:
@kiuby88 Thanks for this. I've left a few minor comments. Not sure why Travis is failing. Any thoughts? It builds fine for me on localhost and Ubuntu/CentOS VMs.

kiuby88 commented:
Thanks for your replies @sjcorbett  
I have addressed some comments. I have to add some unit tests.

I do not know why travis is not happy. I have created a Ubuntu 14.0 VM on AWS to run this branch and maven run all the tests correctly. Let me figure out a bit more about that.

sjcorbett commented:
@kiuby88 Could you rebase this branch on master? I've pushed a couple of commits that may be useful.

kiuby88 commented:
Sure. I saw the PR. I am waiting to your PR will be merged to rebase this PR.

sjcorbett commented:
@kiuby88 I can reproduce this test failure by building it in an Ubuntu 12.04 VM. If you want to try:

``` ruby
# Vagrantfile API/syntax version. Don't touch unless you know what you're doing!
VAGRANTFILE_API_VERSION = "2"

Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
  config.vm.box = "hashicorp/precise64"
  config.vm.network "private_network", ip: "192.168.33.234"

  config.vm.provider "virtualbox" do |vb|
    vb.customize ["modifyvm", :id, "--memory", "2512"]
  end
end
```

I used an Oracle 1.7 JDK and Maven 3.2.5.

sjcorbett commented:
@kiuby88 I think that the test fails because the effect of your changes to `ConfigKeyModifier` and `RelationshipModifier` is dependent on the order in which they are applied when creating the spec. No ordering is guaranteed.

kiuby88 commented:
@sjcorbett  My apologies for leaving this for a week, currently I am swamped. I thought  about the order and so I tried use the same operation (which does not replace values) [here](https://github.com/cloudsoft/brooklyn-tosca/pull/103/commits/9271e664eb5a08bebcd067923510adb19ed09d91), but I was not lucky. I would try the code in Ubuntu 12.04 as you said last week

kiuby88 commented:
Hi @sjcorbett. I think the issue was solved

sjcorbett commented:
@kiuby88 Thanks for revisiting this.

kiuby88 commented:
Thanks for the help @sjcorbett
